From 97cd832babee54bccbd8adc99e9e1a1b28c5ad49 Mon Sep 17 00:00:00 2001
From: Brandon Maier <brandon.maier@gmail.com>
Date: Thu, 19 Dec 2024 21:14:24 -0600
Subject: [PATCH aerc] config: fix system paths when relocating aerc

The following error occurs when packaging aerc for Conda:

 $ aerc
 /home/conda/.config/aerc/aerc.conf not found, installing the system default
 error: open /usr/share/aerc/aerc.conf: no such file or directory

Conda installs aerc to a custom $PREFIX, so the config files will not be
available in /usr/share/aerc. The above error occurs because aerc
searches multiple paths for config paths and /usr/share/aerc happens to
be the last path. The $PREFIX is first in the search list, so we would
expect it to find the Conda installed files.

By dumping out the contents of the search paths, I observed the $PREFIX
paths are being searched but they are corrupted because their strings
end with embedded NULLs (\x00).

Conda builds aerc with a $PREFIX that is extremely long. Then when
installing aerc, it does relocation by replacing the build $PREFIX with
the final install $PREFIX, appending NULLs to fill out the string.

go-lang does not use NULL-terminated strings, so these NULLs are
preserved in the path. Trim any NULLs from the end of $PREFIX paths.

Signed-off-by: Brandon Maier <brandon.maier@gmail.com>
Acked-by: Robin Jarry <robin@jarry.cc>
Upstream: https://git.sr.ht/~rjarry/aerc/commit/d5c04d3edf4824c053f14bbe8092cd0c97a75aad
---
 config/config.go | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/config/config.go b/config/config.go
index 61ac041e..c6713fd4 100644
--- a/config/config.go
+++ b/config/config.go
@@ -33,6 +33,10 @@ func buildDefaultDirs() []string {
 		}
 	}
 
+	// Trim null chars inserted post-build by systems like Conda
+	shareDir := strings.TrimRight(shareDir, "\x00")
+	libexecDir := strings.TrimRight(libexecDir, "\x00")
+
 	// Add custom buildtime dirs
 	if libexecDir != "" && libexecDir != "/usr/local/libexec/aerc" {
 		defaultDirs = append(defaultDirs, xdg.ExpandHome(libexecDir))
-- 
2.47.1

